src/Sarif/Visitors/PerRunPerTargetPerRuleSplittingVisitor.cs

```diff
@@ -10,38 +10,42 @@ public class PerRunPerTargetPerRuleSplittingVisitor : SplittingVisitor
     {
         private static ArtifactLocation s_emptyArtifactLocation = new ArtifactLocation();
 
-        private Dictionary<ArtifactLocation, Dictionary<string, SarifLog>> _targetToRuleMap;
+        private Dictionary<string, Dictionary<string, SarifLog>> _targetToRuleMap;
 
         public PerRunPerTargetPerRuleSplittingVisitor(Func<Result, bool> filteringStrategy = null) : base(filteringStrategy)
         {
         }
 
         public override Run VisitRun(Run node)
         {
-            _targetToRuleMap = new Dictionary<ArtifactLocation, Dictionary<string, SarifLog>>(ArtifactLocation.ValueComparer);
+            _targetToRuleMap = new Dictionary<string, Dictionary<string, SarifLog>>();
             return base.VisitRun(node);
         }
 
         public override Result VisitResult(Result node)
         {
             if (!FilteringStrategy(node)) { return node; }
 
-            string ruleId = node.RuleId;
-
-            if (string.IsNullOrEmpty(ruleId) && node.RuleIndex > -1)
+            string ruleId;
+            if (node.RuleIndex > -1 && CurrentRun.Tool.Driver.Rules.Count > node.RuleIndex)
             {
                 ruleId = CurrentRun.Tool.Driver.Rules?[node.RuleIndex].Id;
             }
+            else
+            {
+                // Remove the rule pattern index from the rule id. e.g. CSCAN0230/5
+                ruleId = node.RuleId.Substring(0, node.RuleId.LastIndexOf('/'));
+            }
 
             ArtifactLocation artifactLocation = s_emptyArtifactLocation;
             if (node.Locations[0].PhysicalLocation?.ArtifactLocation != null)
             {
                 artifactLocation = node.Locations[0].PhysicalLocation?.ArtifactLocation;
             }
 
-            if (!_targetToRuleMap.TryGetValue(artifactLocation, out Dictionary< string, SarifLog > ruleToSarifLogMap))
+            if (!_targetToRuleMap.TryGetValue(artifactLocation.Uri.ToString(), out Dictionary<string, SarifLog> ruleToSarifLogMap))
```

Comment:
Does the default s_emptyArtifactLocation have a populated Uri, if not protect against null ToString call.